bindings/resolver.go

```diff
@@ -0,0 +1,134 @@
+package bindings
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// Binding represents metadata related to an external service.
+type Binding struct {
+
+	// Name is the name of the binding, given by the binding directory name.
+	Name string
+
+	// Path is the path to the binding directory.
+	Path string
+
+	// Type is the type of the binding, given by the content of the 'type' file within the binding directory.
+	Type string
+
+	// Provider is the provider of the binding, given by the content of the 'provider' file within the binding
+	// directory.
+	Provider string
+
+	// Secret is the primary content of the binding. Keys are given by each file name within the binding directory
+	// (other than 'type' or 'provider'), and corresponding values are given by the content of each file.
+	Secret map[string][]byte // TODO: place in custom buffer class
+}
+
+// Resolver resolves service bindings according to the kubernetes binding spec:
+// https://github.com/k8s-service-bindings/spec#workload-projection.
+type Resolver struct {
+	bindingRoot string
+	bindings    []Binding
+}
+
+// NewResolver returns a new service binding resolver. If the SERVICE_BINDING_ROOT environment variable is not set, uses
+// the provided platform directory to resolve bindings at `<platformDir>/bindings`.
+func NewResolver(platformDir string) *Resolver {
+	root := os.Getenv("SERVICE_BINDING_ROOT")
+	if root == "" {
+		root = filepath.Join(platformDir, "bindings")
+	}
+
+	return &Resolver{
+		bindingRoot: root,
+	}
+}
+
+// Resolve returns all bindings matching the given type and optional provider (case-insensitive). To match on type only,
+// provider may be an empty string. Returns an error if there are problems loading bindings from the file system.
+func (r *Resolver) Resolve(typ string, provider string) ([]Binding, error) {
+	if r.bindings == nil {
+		bindings, err := loadBindings(r.bindingRoot)
+		if err != nil {
+			return nil, errors.Wrapf(err, "loading bindings from '%s'", r.bindingRoot)
+		}
+		r.bindings = bindings
+	}
+
+	var resolved []Binding
+	for _, bind := range r.bindings {
+		if (strings.ToLower(bind.Type) == strings.ToLower(typ)) &&
+			(provider == "" || strings.ToLower(bind.Provider) == strings.ToLower(provider)) {
+			resolved = append(resolved, bind)
+		}
+	}
+	return resolved, nil
+}
+
+// ResolveOne returns a single binding matching the given type and optional provider (case-insensitive). To match on
+// type only, provider may be an empty string. Returns an error if the number of matched bindings is not exactly one, or
+// if there are problems loading bindings from the file system.
+func (r *Resolver) ResolveOne(typ string, provider string) (Binding, error) {
```

Comment:
@sophiewigmore I debated this too. I kept it because it's implemented similarly in `libpak`, and I also saw at least one buildpack implementation use it (can't remember off-hand). So I felt like it might be good as an option. I was also planning to use it in the dependency mapping logic here in `packit` but it turned out not to need it.

Reply:
I could see it being useful when a buildpack expects only one binding for a given type and wouldn't be able to decide what to do if it were ambiguous anyway.